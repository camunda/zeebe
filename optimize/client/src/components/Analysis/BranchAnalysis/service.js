/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {post} from 'request';

export async function loadFrequencyData(
  processDefinitionKey,
  processDefinitionVersions,
  tenantIds,
  identifier,
  filter
) {
  const response = await post(
    'api/report/evaluate',
    createFlowNodeFrequencyReport(
      processDefinitionKey,
      processDefinitionVersions,
      tenantIds,
      identifier,
      filter
    )
  );

  return await response.json();
}

function createFlowNodeFrequencyReport(key, versions, tenantIds, identifier, filter) {
  return {
    combined: false,
    reportType: 'process',
    data: {
      definitions: [{key, versions, tenantIds, identifier}],
      filter,
      view: {
        entity: 'flowNode',
        properties: ['frequency'],
      },
      groupBy: {
        type: 'flowNodes',
        unit: null,
      },
      visualization: 'heat',
    },
  };
}

export async function loadCorrelationData(
  processDefinitionKey,
  processDefinitionVersions,
  tenantIds,
  filter,
  gateway,
  end
) {
  const response = await post('api/analysis/correlation', {
    processDefinitionKey,
    processDefinitionVersions,
    tenantIds,
    filter,
    gateway,
    end,
  });

  return await response.json();
}
