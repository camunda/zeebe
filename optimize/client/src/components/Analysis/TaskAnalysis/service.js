/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {post} from 'request';

export async function loadNodesOutliers(config) {
  const response = await post('api/analysis/flowNodeOutliers', config);
  return await response.json();
}

export {
  loadDurationData,
  loadCommonOutliersVariables,
  getOutlierSummary,
  shouldUseLogharitmicScale,
} from './service.ts';
