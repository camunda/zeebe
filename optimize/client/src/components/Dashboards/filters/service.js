/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {post, get} from 'request';

export async function getVariableNames(reportIds) {
  const response = await post('api/variables/reports', {reportIds});

  return await response.json();
}

export async function getVariableValues(reportIds, name, type, numResults, valueFilter) {
  const response = await post('api/variables/values/reports', {
    reportIds,
    name,
    type,
    valueFilter,
    numResults,
    resultOffset: 0,
  });

  return await response.json();
}

export async function loadUsersByReportIds(type, payload) {
  const response = await post(`api/${type}/search/reports`, payload);

  return await response.json();
}

export async function getAssigneeNames(type, listOfIds) {
  const response = await get('api/' + type, {idIn: listOfIds.join(',')});

  return await response.json();
}

export function isOfType(filter, availableFilter) {
  switch (filter.type) {
    case 'runningInstancesOnly':
    case 'completedInstancesOnly':
    case 'canceledInstancesOnly':
    case 'nonCanceledInstancesOnly':
    case 'suspendedInstancesOnly':
    case 'nonSuspendedInstancesOnly':
      return availableFilter.type === 'state';
    case 'instanceStartDate':
    case 'instanceEndDate':
    case 'assignee':
    case 'candidateGroup':
      return availableFilter.type === filter.type;
    case 'variable':
      return (
        availableFilter.type === 'variable' &&
        availableFilter.data.type === filter.data.type &&
        availableFilter.data.name === filter.data.name
      );
    default:
      return false;
  }
}
