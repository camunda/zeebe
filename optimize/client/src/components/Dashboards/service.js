/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {get, del, post} from 'request';

export async function shareDashboard(dashboardId) {
  const body = {
    dashboardId,
  };
  const response = await post(`api/share/dashboard`, body);

  const json = await response.json();
  return json.id;
}

export async function getSharedDashboard(reportId) {
  const response = await get(`api/share/dashboard/${reportId}`);

  if (response.status > 201) {
    return '';
  } else {
    const json = await response.json();
    return json.id;
  }
}

export async function revokeDashboardSharing(id) {
  return await del(`api/share/dashboard/${id}`);
}

export async function isAuthorizedToShareDashboard(dashboardId) {
  try {
    const response = await get(`api/share/dashboard/${dashboardId}/isAuthorizedToShare`);
    return response.status === 200;
  } catch (error) {
    return false;
  }
}

export function convertFilterToDefaultValues(availableFilter, filters) {
  // dates
  if (availableFilter.type === 'instanceStartDate' || availableFilter.type === 'instanceEndDate') {
    return filters.find(({type}) => type === availableFilter.type)?.data ?? null;
  }

  // assignee / candidate group
  if (availableFilter.type === 'assignee' || availableFilter.type === 'candidateGroup') {
    return (
      filters.find(
        ({type, data}) =>
          type === availableFilter.type && data.operator === availableFilter.data.operator
      )?.data.values ?? null
    );
  }

  // state
  if (availableFilter.type === 'state') {
    const defaultValues = filters
      .filter(({type}) =>
        [
          'runningInstancesOnly',
          'completedInstancesOnly',
          'canceledInstancesOnly',
          'nonCanceledInstancesOnly',
          'suspendedInstancesOnly',
          'nonSuspendedInstancesOnly',
        ].includes(type)
      )
      .map(({type}) => type);

    if (defaultValues.length > 0) {
      return defaultValues;
    }
    return null;
  }

  // Boolean Variables
  if (availableFilter.type === 'variable' && availableFilter.data.type === 'Boolean') {
    return (
      filters.find(
        ({type, data}) =>
          type === 'variable' && data.type === 'Boolean' && data.name === availableFilter.data.name
      )?.data.data.values ?? null
    );
  }

  // Date Variables
  if (availableFilter.type === 'variable' && availableFilter.data.type === 'Date') {
    return (
      filters.find(
        ({type, data}) =>
          type === 'variable' && data.type === 'Date' && data.name === availableFilter.data.name
      )?.data.data ?? null
    );
  }

  // string and number variables
  if (availableFilter.type === 'variable') {
    return (
      filters.find(
        ({type, data}) =>
          type === 'variable' &&
          data.type === availableFilter.data.type &&
          data.name === availableFilter.data.name &&
          data.data.operator === availableFilter.data.data.operator
      )?.data.data.values ?? null
    );
  }

  return null;
}

export function getDefaultFilter(availableFilters) {
  const filters = [];

  availableFilters.forEach((availableFilter) => {
    if (availableFilter?.data?.defaultValues) {
      if (availableFilter.type === 'state') {
        return filters.push(
          ...availableFilter.data.defaultValues.map((type) => ({
            filterLevel: 'instance',
            type,
            data: null,
          }))
        );
      }

      if (
        availableFilter.type === 'instanceStartDate' ||
        availableFilter.type === 'instanceEndDate'
      ) {
        return filters.push({
          type: availableFilter.type,
          filterLevel: 'instance',
          data: availableFilter.data.defaultValues,
        });
      }

      if (availableFilter.type === 'assignee' || availableFilter.type === 'candidateGroup') {
        return filters.push({
          type: availableFilter.type,
          filterLevel: 'view',
          data: {
            operator: availableFilter.data.operator,
            values: availableFilter.data.defaultValues,
          },
        });
      }

      if (availableFilter.type === 'variable' && availableFilter.data.type === 'Boolean') {
        return filters.push({
          type: availableFilter.type,
          filterLevel: 'instance',
          data: {
            name: availableFilter.data.name,
            type: availableFilter.data.type,
            data: {values: availableFilter.data.defaultValues},
          },
        });
      }

      if (availableFilter.type === 'variable' && availableFilter.data.type === 'Date') {
        return filters.push({
          type: availableFilter.type,
          filterLevel: 'instance',
          data: {
            name: availableFilter.data.name,
            type: availableFilter.data.type,
            data: availableFilter.data.defaultValues,
          },
        });
      }

      if (availableFilter.type === 'variable') {
        return filters.push({
          type: availableFilter.type,
          filterLevel: 'instance',
          data: {
            name: availableFilter.data.name,
            type: availableFilter.data.type,
            data: {
              operator: availableFilter.data.data.operator,
              values: availableFilter.data.defaultValues,
            },
          },
        });
      }
    }
  });

  return filters;
}
