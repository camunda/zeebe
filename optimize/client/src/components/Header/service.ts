/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {get} from 'request';

export async function isEventBasedProcessEnabled(): Promise<boolean> {
  const response = await get('api/eventBasedProcess/isEnabled');
  const value = await response.text();

  return value === 'true';
}

export async function getUserToken(): Promise<string> {
  const response = await get('api/token');
  const value = (await response.json()) as {token: string};
  return value.token;
}
