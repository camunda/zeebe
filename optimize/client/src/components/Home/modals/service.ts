/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {get} from 'request';
import {Definition, Tenant} from 'types';
import {formatters, UNAUTHORIZED_TENANT_ID} from 'services';
import {t} from 'translation';

const {formatTenantName} = formatters;

export type TenantWithDefinitions = Tenant & {definitions: Definition};

export type DefinitionWithTenants = Omit<Definition, 'key'> & {tenants: Tenant[]; key: string};

export async function getDefinitionsWithTenants(): Promise<DefinitionWithTenants[]> {
  const response = await get('api/definition');

  return await response.json();
}

export async function getTenantsWithDefinitions(): Promise<TenantWithDefinitions[]> {
  const response = await get('api/definition/_groupByTenant');

  return await response.json();
}

export async function getDefinitionTenants(
  defintionKey: string,
  defintionType: string
): Promise<DefinitionWithTenants> {
  const response = await get(`api/definition/${defintionType}/${defintionKey}`);

  return await response.json();
}

export function formatTenants(tenants: Tenant[], selectedTenants: Tenant[]) {
  return tenants.map(({id, name}, index) => {
    if (id === UNAUTHORIZED_TENANT_ID) {
      return {
        id: index,
        label: t('home.sources.unauthorizedTenant'),
        checked: true,
        disabled: true,
      };
    }

    return {
      id,
      label: formatTenantName({id, name}),
      checked: selectedTenants.some((tenant) => tenant.id === id),
    };
  });
}
