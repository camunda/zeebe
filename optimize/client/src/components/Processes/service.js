/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {get, put} from 'request';

import {formatters} from 'services';

export async function loadProcesses(sortBy, sortOrder) {
  const params = {};
  if (sortBy && sortOrder) {
    params.sortBy = sortBy;
    params.sortOrder = sortOrder;
  }

  const response = await get('api/process/overview', params);
  return await response.json();
}

export function updateProcess(processDefinitionKey, payload) {
  return put(`api/process/${processDefinitionKey}`, payload);
}

export async function loadManagementDashboard() {
  const response = await get(`api/dashboard/management`);

  return await response.json();
}

export function isSuccessful({target, unit, value, isBelow, measure}) {
  const actualValue = Number(value);
  const targetMs =
    measure === 'duration' && unit
      ? formatters.convertToMilliseconds(target, unit)
      : Number(target);

  return isBelow ? targetMs > actualValue : targetMs < actualValue;
}
