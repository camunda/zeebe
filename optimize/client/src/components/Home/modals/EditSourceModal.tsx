/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {useCallback, useEffect, useState} from 'react';
import {Button} from '@carbon/react';

import {Modal, Checklist} from 'components';
import {t} from 'translation';
import {Tenant} from 'types';
import {useErrorHandling} from 'hooks';
import {UNAUTHORIZED_TENANT_ID} from 'services';

import {getDefinitionTenants, formatTenants} from './service';

interface Source {
  tenants: Tenant[];
  definitionKey: string;
  definitionType: string;
  definitionName: string;
}

interface EditSourceModalProps {
  source: Source;
  onConfirm: (ids: (string | null)[]) => void;
  onClose: () => void;
}

export default function EditSourceModal({source, onConfirm, onClose}: EditSourceModalProps) {
  const [selectedTenants, setSelectedTenants] = useState(source.tenants);
  const [definitionTenants, setDefinitionTenants] = useState<Tenant[]>();
  const {mightFail} = useErrorHandling();

  const getUnauthorizedTenants = useCallback(
    (tenants: Tenant[]) => tenants.filter((tenant) => tenant.id === UNAUTHORIZED_TENANT_ID),
    []
  );

  useEffect(() => {
    const {definitionKey, definitionType} = source;
    mightFail(getDefinitionTenants(definitionKey, definitionType), ({tenants}) => {
      setDefinitionTenants([...tenants, ...getUnauthorizedTenants(source.tenants)]);
    });
  }, [source, mightFail, getUnauthorizedTenants]);

  const onConfirmHandler = () => {
    if (selectedTenants.length > 0) {
      onConfirm(selectedTenants.map(({id}) => id));
    }
  };

  const updateSelectedTenants = (newSelectedTenants: (Tenant | undefined)[]) => {
    if (!newSelectedTenants.length) {
      setSelectedTenants((selectedTenants) => getUnauthorizedTenants(selectedTenants));
    } else {
      setSelectedTenants(
        newSelectedTenants.filter((tenant): tenant is Tenant => typeof tenant !== 'undefined')
      );
    }
  };

  const modalTitle = `${t('common.editTenants')} - ${source.definitionName || source.definitionKey}`;

  return (
    <Modal className="EditSourceModal" open onClose={onClose}>
      <Modal.Header title={modalTitle} />
      <Modal.Content>
        <Checklist
          columnLabel={t('common.tenant.label')}
          selectedItems={selectedTenants}
          allItems={definitionTenants ?? []}
          onChange={updateSelectedTenants}
          formatter={formatTenants}
          loading={!definitionTenants}
        />
      </Modal.Content>
      <Modal.Footer>
        <Button kind="secondary" className="cancel" onClick={onClose}>
          {t('common.cancel')}
        </Button>
        <Button className="confirm" disabled={!selectedTenants.length} onClick={onConfirmHandler}>
          {t('common.apply')}
        </Button>
      </Modal.Footer>
    </Modal>
  );
}
