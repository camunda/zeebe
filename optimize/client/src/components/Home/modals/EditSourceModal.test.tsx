/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {runAllEffects} from '__mocks__/react';
import {ComponentProps} from 'react';
import {shallow} from 'enzyme';

import {UNAUTHORIZED_TENANT_ID} from 'services';

import EditSourceModal from './EditSourceModal';
import {getDefinitionTenants} from './service';

jest.mock('./service', () => ({getDefinitionTenants: jest.fn()}));
jest.mock('hooks', () => ({
  ...jest.requireActual('hooks'),
  useErrorHandling: () => ({
    mightFail: jest.fn().mockImplementation((data, cb) => cb(data)),
  }),
}));

const props: ComponentProps<typeof EditSourceModal> = {
  onClose: jest.fn(),
  onConfirm: jest.fn(),
  source: {
    definitionKey: 'defKey',
    definitionName: 'defName',
    definitionType: 'process',
    tenants: [
      {id: null, name: 'Not defined'},
      {id: UNAUTHORIZED_TENANT_ID, name: 'unauthorizedTenant'},
    ],
  },
};

(getDefinitionTenants as jest.Mock).mockReturnValue({
  ...props.source,
  tenants: [
    {id: null, name: 'Not defined'},
    {id: 'test', name: 'testName'},
  ],
});

it('should match snapshot', () => {
  const node = shallow(<EditSourceModal {...props} />);

  runAllEffects();

  expect(node).toMatchSnapshot();
});

it('should get defintion tenants on mount', () => {
  shallow(<EditSourceModal {...props} />);

  runAllEffects();

  expect(getDefinitionTenants).toHaveBeenCalled();
});

it('should update selected tenants on itemList change', () => {
  const node = shallow(<EditSourceModal {...props} />);

  node.find('Checklist').simulate('change', [{id: null, name: 'Not defined'}]);
  node.find('.confirm').simulate('click');

  expect(props.onConfirm).toHaveBeenCalledWith([null]);
});

it('should not deselect unauthorized tenants', () => {
  const node = shallow(<EditSourceModal {...props} />);

  node.find('Checklist').simulate('change', []);

  expect(node.find('Checklist').prop('selectedItems')).toEqual([
    {id: UNAUTHORIZED_TENANT_ID, name: 'unauthorizedTenant'},
  ]);
});
