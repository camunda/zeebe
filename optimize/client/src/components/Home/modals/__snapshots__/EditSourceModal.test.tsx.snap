// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<Modal
  className="EditSourceModal"
  onClose={[MockFunction]}
  open={true}
>
  <ForwardRef(ModalHeader)
    title="Edit tenants - defName"
  />
  <ForwardRef>
    <Checklist
      allItems={
        Array [
          Object {
            "id": null,
            "name": "Not defined",
          },
          Object {
            "id": "test",
            "name": "testName",
          },
          Object {
            "id": "__unauthorizedTenantId__",
            "name": "unauthorizedTenant",
          },
        ]
      }
      columnLabel="Tenant"
      loading={false}
      onChange={[Function]}
      selectedItems={
        Array [
          Object {
            "id": null,
            "name": "Not defined",
          },
          Object {
            "id": "__unauthorizedTenantId__",
            "name": "unauthorizedTenant",
          },
        ]
      }
    />
  </ForwardRef>
  <ForwardRef(ModalFooter)>
    <Button
      className="cancel"
      kind="secondary"
      onClick={[MockFunction]}
    >
      Cancel
    </Button>
    <Button
      className="confirm"
      disabled={false}
      onClick={[Function]}
    >
      Apply
    </Button>
  </ForwardRef(ModalFooter)>
</Modal>
`;
